src/magicbox/QueriesProcessor.ts

```diff
@@ -0,0 +1,91 @@
+import { defaults } from 'underscore';
+
+export interface IQueriesProcessorOptions {
+  timeout: number;
+}
+
+export enum QueryProcessResultStatus {
+  Finished = 'finished',
+  TimedOut = 'timedout',
+  Overriden = 'overriden'
+}
+
+export type QueryProcessResult<ItemType> =
+  | {
+      status: QueryProcessResultStatus.Finished | QueryProcessResultStatus.TimedOut;
+      items: ItemType[];
+    }
+  | {
+      status: QueryProcessResultStatus.Overriden;
+    };
+
+export class QueriesProcessor<ItemType> {
+  private static async streamQueries<ItemType>(queries: Promise<ItemType[]>[], output: ItemType[]) {
+    await Promise.all(queries.map(query => query.then(items => output.push(...items))));
```

Comment:
> I prefer to do the `then` after the `Promise.all` resolves to reduce the nesting and better leverage `async`/`await`.

Unfortunately, it currently looks like that because it's meant to be used by `processQueries` in such a way that it's possible to get the resolved output thus far before all of them have resolved.

This is so that `processQueries` returns the same result as `SuggestionsManager.mergeSuggestions`.

> I also think it would be nicer to not pass the second arg but rather change the function from `void` to returning the result.

Unfortunately, returning the result would prevent the behaviour I explained above.

> For the name, maybe `resolveQueriesAndConcatenateResults`?

This is a much less confusing name than `streamQueries`, but should I use it given the behaviour I explained above?

> Also, since the method is called by an instance method, it can be an instance too instead of `static`.

Good point, I'll change it to be an instance method.

Reply:
Why is it important to get the resolved output before all of the promises have resolved?

Reply:
Because this component is going to be used by `SuggestionsManager` to replace `mergeSuggestions`. As of right now, `SuggestionsManager` will, when timing out, return the results of promises resolved so far.